internal/datastore/crdb/tx.go

```diff
@@ -37,36 +50,73 @@ type transactionFn func(tx pgx.Tx) error
 
 type executeTxRetryFunc func(context.Context, conn, pgx.TxOptions, transactionFn) error
 
+type RetryError struct {
+	err error
+}
+
+func (e RetryError) Error() string {
+	return fmt.Errorf("unable to retry conflicted transaction: %w", e.err).Error()
+}
+
+func (e RetryError) Unwrap() error {
+	return e.err
+}
+
 func executeWithMaxRetries(max int) executeTxRetryFunc {
 	return func(ctx context.Context, conn conn, txOptions pgx.TxOptions, fn transactionFn) (err error) {
-		return execute(ctx, conn, txOptions, fn, max)
+		return executeWithResets(ctx, conn, txOptions, fn, max)
 	}
 }
 
-// adapted from https://github.com/cockroachdb/cockroach-go
-func execute(ctx context.Context, conn conn, txOptions pgx.TxOptions, fn transactionFn, maxRetries int) (err error) {
+// executeWithResets executes transactionFn and resets the tx when ambiguous crdb errors are encountered.
+func executeWithResets(ctx context.Context, conn conn, txOptions pgx.TxOptions, fn transactionFn, maxRetries int) (err error) {
+	var retries, resets int
+	defer func() {
+		retryHistogram.Observe(float64(retries))
+		resetHistogram.Observe(float64(resets))
+	}()
+
 	var tx pgx.Tx
-	tx, err = conn.BeginTx(ctx, txOptions)
-	if err != nil {
-		return err
-	}
 	defer func() {
 		if err == nil {
-			_ = tx.Commit(ctx)
-			return
+			if commitErr := tx.Commit(ctx); commitErr == nil {
+				return
+			}
 		}
 		_ = tx.Rollback(ctx)
 	}()
 
-	if _, err = tx.Exec(ctx, "SAVEPOINT cockroach_restart"); err != nil {
+	// n maxRetries can yield n+1 executions of the transaction fn
+	for retries = 0; retries <= maxRetries; retries++ {
+		tx, err = resetExecution(ctx, conn, tx, txOptions)
+		if err != nil {
+			log.Ctx(ctx).Err(err).Msg("error resetting transaction")
+			resets++
+			continue
+		}
+
+		var retryAttempts int
+		retryAttempts, err = executeWithRetries(ctx, tx, fn, maxRetries-retries)
```

Comment:
Maybe use a newly named variable for err, and then you can use `:=` here?

Reply:
Using the named return var name so that `err` can be used below

Reply:
Yeah, I prefer in these cases to use a newly named error var, to remove the need for the `var`